### ISSUE TITLE
Handling missing fields

### ISSUE THREAD

**Issue by glasserc** [Kinto/kinto#1257]
If you try to filter for fields whose value is the empty string, it also matches records without those fields:

```
> http 'https://kinto.dev.mozaws.net/v1/buckets/46524be8-0ad7-3ac6-e260-71f8993feffa/collections/tasks/records?any_arbitrary_field=""' -a user:pass
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Next-Page, Pragma, Retry-After, Backoff, Alert, Cache-Control, ETag, Last-Modified, Expires, Total-Records, Content-Length
Cache-Control: no-cache, no-store
Connection: keep-alive
Content-Length: 429
Content-Type: application/json
Date: Tue, 06 Jun 2017 20:50:47 GMT
ETag: "1496685126658"
Last-Modified: Mon, 05 Jun 2017 17:52:06 GMT
Server: nginx
Total-Records: 4
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "done": false,
            "id": "3dd095cb-f4f7-4c87-8112-6cd3e7660f41",
            "last_modified": 1496685088616,
            "title": "test"
        },
        {
            "done": false,
            "id": "0717608f-1cf2-472b-8a49-3136d2811946",
            "last_modified": 1496685088608,
            "title": "blbl"
        },
        {
            "done": false,
            "id": "65e5b714-dcb2-4136-836c-996e83514297",
            "last_modified": 1496524477816,
            "title": "to a poo"
        },
        {
            "done": false,
            "id": "5a1fc9a7-6c6a-428e-b932-930a69d6495a",
            "last_modified": 1496422987261,
            "title": "444"
        }
    ]
}
```

This is because of https://github.com/Kinto/kinto/blob/master/kinto/core/storage/postgresql/__init__.py#L709. Postgres appears to be the reference implementation of Kinto's storage backend (see #815), so you might expect this behavior to leak into other backends too.

The [documentation](https://github.com/Kinto/kinto/blob/master/docs/api/1.x/filtering.rst) is pretty scanty on the subject of missing fields, saying only `Will return an error if a field is unknown`, and I think that's meant to only apply to `like_`.

I'm not 100% sure what we should expect this to do (much less how to implement it), but it seems like you'd expect missing fields to compare as different from everything, including null (see #1215), and not match `ILIKE`. In the tests, for instance, there's a couple tests that filter for `deleted != true` (which is intended to match "live" records) or `deleted == false` (which is intended not to match any records).

**Comment by glasserc** [comment]
There's also the order of sort order. Postgres documentation says that `By default, null values sort as if larger than any non-null value; that is, NULLS FIRST is the default for DESC order, and NULLS LAST otherwise.` However, because of our strange coalescing behavior, you can get it to do strange things:

```
> http -a 'user:pass' POST localhost:8888/v1/buckets/default/collections/icecream/records data:='{"flavor": "strawberry"}'
> http -a 'user:pass' POST localhost:8888/v1/buckets/default/collections/icecream/records data:='{"flavor": "blueberry","author": "Ethan"}'
> http -a 'user:pass' POST localhost:8888/v1/buckets/default/collections/icecream/records data:='{"flavor": "raspberry","author": null}'
> http -a 'user:pass' 'localhost:8888/v1/buckets/default/collections/icecream/records?_sort=author'
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Total-Records, Cache-Control, Content-Length, Alert, ETag, Expires, Pragma, Last-Modified, Retry-After, Next-Page, Backoff
Cache-Control: no-cache, no-store
Content-Length: 333
Content-Type: application/json
Date: Tue, 06 Jun 2017 21:17:07 GMT
Etag: "1496783699002"
Last-Modified: Tue, 06 Jun 2017 21:14:59 GMT
Server: waitress
Total-Records: 3
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "author": null,
            "flavor": "raspberry",
            "id": "265506f9-2a07-4ba4-a548-8b7cd3b9a457",
            "last_modified": 1496783699002
        },
        {
            "author": "Ethan",
            "flavor": "blueberry",
            "id": "0a1b92fd-ddf7-460b-aceb-59b5f599c0eb",
            "last_modified": 1496783688505
        },
        {
            "flavor": "strawberry",
            "id": "480f82db-1dc9-4a61-ac02-fefcda8c8928",
            "last_modified": 1496783673440
        }
    ]
}

# Seems legit. Let's paginate
> http -a 'user:pass' 'localhost:8888/v1/buckets/default/collections/icecream/records?_sort=author&gt_author=a'
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Total-Records, Cache-Control, Content-Length, Alert, ETag, Expires, Pragma, Last-Modified, Retry-After, Next-Page, Backoff
Cache-Control: no-cache, no-store
Content-Length: 124
Content-Type: application/json
Date: Tue, 06 Jun 2017 21:17:41 GMT
Etag: "1496783699002"
Last-Modified: Tue, 06 Jun 2017 21:14:59 GMT
Server: waitress
Total-Records: 1
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "author": "Ethan",
            "flavor": "blueberry",
            "id": "0a1b92fd-ddf7-460b-aceb-59b5f599c0eb",
            "last_modified": 1496783688505
        }
    ]
}
# Wait... what happened to strawberry?
```

**Comment by leplatrem** [comment]
Related #816  #1157

**Comment by glasserc** [comment]
Interestingly, this behavior seems to be even more bonkers for integer fields:

```
> http -a 'user:pass' 'localhost:8888/v1/buckets/default/collections/icecream2/records'
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Content-Length, Expires, Last-Modified, Total-Records, Retry-After, Pragma, Alert, Cache-Control, ETag, Backoff, Next-Page
Cache-Control: no-cache, no-store
Content-Length: 327
Content-Type: application/json
Date: Wed, 07 Jun 2017 15:55:20 GMT
Etag: "1496850665926"
Last-Modified: Wed, 07 Jun 2017 15:51:05 GMT
Server: waitress
Total-Records: 3
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "author": null,
            "flavor": "raspberry",
            "id": "9b5a421c-8860-4331-8fdb-590c41bb443c",
            "last_modified": 1496850665926
        },
        {
            "author": 1,
            "flavor": "blueberry",
            "id": "23697cff-be13-419e-b6d5-9023b2b09f62",
            "last_modified": 1496850661019
        },
        {
            "flavor": "strawberry",
            "id": "dc70e65a-7835-4b38-a9ca-59de6c7e532a",
            "last_modified": 1496850653779
        }
    ]
}
> http -a 'user:pass' 'localhost:8888/v1/buckets/default/collections/icecream2/records?lt_author=2'
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Content-Length, Expires, Last-Modified, Total-Records, Retry-After, Pragma, Alert, Cache-Control, ETag, Backoff, Next-Page
Cache-Control: no-cache, no-store
Content-Length: 118
Content-Type: application/json
Date: Wed, 07 Jun 2017 15:55:34 GMT
Etag: "1496850665926"
Last-Modified: Wed, 07 Jun 2017 15:51:05 GMT
Server: waitress
Total-Records: 1
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "author": 1,
            "flavor": "blueberry",
            "id": "23697cff-be13-419e-b6d5-9023b2b09f62",
            "last_modified": 1496850661019
        }
    ]
}
# Hmm, OK, I guess null and missing fields should be greater than?...
> http -a 'user:pass' 'localhost:8888/v1/buckets/default/collections/icecream2/records?gt_author=2'
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Content-Length, Expires, Last-Modified, Total-Records, Retry-After, Pragma, Alert, Cache-Control, ETag, Backoff, Next-Page
Cache-Control: no-cache, no-store
Content-Length: 11
Content-Type: application/json
Date: Wed, 07 Jun 2017 15:55:58 GMT
Etag: "1496850665926"
Last-Modified: Wed, 07 Jun 2017 15:51:05 GMT
Server: waitress
Total-Records: 0
X-Content-Type-Options: nosniff

{
    "data": []
}

> http -a 'user:pass' 'localhost:8888/v1/buckets/default/collections/icecream2/records?min_author=2'
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Content-Length, Expires, Last-Modified, Total-Records, Retry-After, Pragma, Alert, Cache-Control, ETag, Backoff, Next-Page
Cache-Control: no-cache, no-store
Content-Length: 11
Content-Type: application/json
Date: Wed, 07 Jun 2017 15:56:05 GMT
Etag: "1496850665926"
Last-Modified: Wed, 07 Jun 2017 15:51:05 GMT
Server: waitress
Total-Records: 0
X-Content-Type-Options: nosniff

{
    "data": []
}
# wat?
```

**Comment by glasserc** [comment]
Fixed by #1258.
